### PR TITLE
Turn the login-fire element into a collection

### LINKED ISSUE #45: [proposal] Turn the element into a collection
Hi!

I wanted just say I'm working on a fork of this element. It was restructured for integrating a shared behaviour and a new `login-fire-button` element that offers login and logout. There is a little work but I'll make a PR when everything will be ready. I hope it will please to you.

### PR DESCRIPTION
**Pull request by masonlouchart**
This pull request contains lot of work and most important to know is it could be a breaking change. The purpose was to be able to synchronize the different parts of the original login-fire by creating a behavior. A new login-fire-button was added, the SVGs are now packaged and the entire source code was reviewed to use those new elements.

The way to define what providers to use changed. I can add a behavior to convert booleans into array of providers. The reason why I changed the way to select providers to be used is for offering the possibility to define the order of the buttons.

Closes #45 

Ref #16, #44

### CODE REVIEW

**Comment by sbeleidy** [comment]
Hi @LM450N, thanks for this PR. Sorry we haven't responded yet. We still haven't had a chance to go through all the changes. If there is a subgroup of changes that you'd like merged right away feel free to make a separate PR for those so we can get that updated soon.

For the rest of the changes, it may take some time for us to get back to you. Thank you so much for the contribution and sorry again for the late response. 

For any discussions, feel free to get in touch on the [gitter chat](https://gitter.im/convoo/login-fire) if you'd like.

**Comment by masonlouchart** [comment]
Hi @sbeleidy !

No problem, I understand. Take your time 😉

**Comment by bkawk** [comment]
I notice the Google SVG is not displaying as expected 

screen shot 2017-01-19 at 1 16 54 pm

**Comment by masonlouchart** [comment]
On Google Chrome the SVG is well displayed at the first time. But, on the `login-fire-social` demo page, I click on the first Google sign-in button the SVG of the second one is like @bkawk shown. If I log-out and log-in with the second button, its icon is working well.

On Firefox the Google SVG icon is not at all displayed. There is definitively an issue with this SVG in particular.

File: login-fire-panel.html [hunk 1]
```diff
@@ -0,0 +1,226 @@
+<!--
+@license
+Copyright (c) 2016 Convoo All Rights Reserved.
+Use of this source code is governed by a MIT license that can be found in the
+LICENSE file or at https://github.com/convoo/login-fire/blob/master/LICENSE.
+-->
+<link rel="import" href="../polymer/polymer.html">
+<link rel="import" href="./login-fire-common-behavior.html">
```
Comment (sbeleidy) [inline login-fire-panel.html:8]: Stay consistent with imports. Use either "./" or direct import as seen below.

File: login-fire-social.html [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+        margin-bottom: 0;
+      }
+    </style>
+
+    <template is="dom-repeat" items="{{_providers}}">
+      <login-fire-button
+        app-name="[[appName]]"
+        provider="[[item.id]]"
+        user="{{_user}}"
+        status-known="{{_statusKnown}}"
```
Comment (sbeleidy) [inline login-fire-social.html:56]: Not sure how _statusKnown and _signedIn variables are used/useful. Could you please explain?

**Comment by sbeleidy** [comment]
Hey @LM450N, thank you for putting in so much work into this! A lot of this stuff looks great! I hope you don't mind but we'll be sending comments as we find issues. We'll try to group them as much as we can and cover everything at once but we'll probably miss something and have to comment about it later. If you are to make any future PRs I would highly recommend a branch for each added feature. It would be much easier to merge and comment on. 

When we merge this it will be part of [login-fire v2](https://github.com/convoo/login-fire/milestone/1) so this may take a while to be pushed out and downloadable via bower.

We will create a v2 branch so you can bower install from there. We will also update this PR to merge into that instead of master. 

Things we really like:

* Implemented behaviors
Love this. Thank you! 
* Providers as a string
In addition to specifying the order, I think this may later on be used to handle [custom authentication providers](https://firebase.google.com/docs/auth/web/custom-auth) #49. 
* Debug console logs
I really like this and it may be part of how we implement #44. We could create a computed debug property that has all the information in it and just console.log it instead of adding dom elements. 
* Shared styles
This is also great, not sure how it affects #7 yet though - might make things easier.
* Auto Sign Up - #30 
Still need to look more into this and test it but it looks like a great start.
* Event handling
The way the events are done and documented is great. We may change the names later as well #50. 
* Getting rid of useless tests
Thanks! Given the way you updated things though we may also be able to start adding test. Added #51
* Customizable messages
Cool, we should also see if anything else should be customizable.

Issues we have found so far:

- [x] Buttons switching to logout. 
I don't like this feature and would prefer to go with something like #41 instead. Could you give us an example of why this would be better? I'm more used to sites having login and logout in two separate areas. Logout is usually behind a dropdown but sometimes can be somewhere as a button. It is also likely placed somewhere else since when you sign in, you usually navigate to a different page. This also feels like it would only be useful if you have only one provider. If we were to keep this, I think it would not be on by default and we would add a boolean flag for it.
The only other way I see this being useful is if you're connecting an account to multiple providers and the user wants to disconnect one. This would require implementing #28. 
- [x] Anonymous mini flat button not showing.
screen shot 2017-01-21 at 12 22 56 am
- [x] Rename login-fire-panel back to login-fire
I am fine with email being renamed form and with the names appearing after login-fire but the "panel" is the core point of login-fire and should stay named as login-fire. Is there a reason you think it should stay as panel? If the reason is imports, the panel should import all the elements since it likely needs them.
-[ ] icons file
There was a reason we switched the icons to inline. Having an icons file is nicer but I believe was an issue due to polymer build. If that is fixed, we can bring it back as a file. See #33.

Some other comments have been made on the code. 

This is what we have for now. Looking forward to discussing all this and hopefully merging it soon. Thank you again for all the effort you put into this!

File: login-fire-panel.html (hunk 1 above)
Comment (masonlouchart) [reply login-fire-panel.html:8]: You're right. Little mistake I missed. I fixed it.

File: login-fire-social.html (hunk 2 above)
Comment (masonlouchart) [reply login-fire-social.html:56]: Those attributes are used by each `login-fire-button` insert into the DOM. It's the simplest way to synchronize them.

**Comment by masonlouchart** [comment]
Thank you for your feedback on the PR. I really appreciate that! Sorry for this bunch of code. I was wanting to insert behavior and a new component and it seemed complicated to make them cohabit without almost duplicate all elements. But you're right it would be easier to review it.

- Buttons switching to logout:
I also planned to add an argument to the `login-fire-button` for turning it into a generic logout button. For example, you could put the `login-fire-button` inside another page. The user can sign-in anywhere with any provider, the `fire-login-button` with the good attribute will take the style according to the provider used and will automatically able to log out the user because of the behavior inheritance.

- Anonymous mini flat button not showing
It's the button shown to the right. It's the button shown to the right. In fact, there is only three buttons in the demo page fo this one because it cannot sign-in with redirection.

- Rename login-fire-panel back to login-fire
This change came from the process of the refactoring. Since it was updated to use the other components you can rename it back to `login-fire`. And I vote for it!

The shared style and separated file for SVG icons were ones of the first things I changed. Now, all elements are using the `login-fire-button` by composition. I don't see any advantage to keep the style and icons shared anymore except readability. Moreover, as you pointed, it can be source of issue. I have to test it, but it could be the source of the issue about the Google icon displaying.

I also want to add unit tests but you know, I don't want to put more code on this PR (except for fixes). I can't imagine the nightmare for you to review this work and for this, I present to you my apologies.

File: login-fire-social.html (hunk 2 above)
Comment (sbeleidy) [reply login-fire-social.html:56]: status-known is read only and I'm not really sure how that helps with signed-in. Does it mean that if you sign in to one you can sign out from all the other buttons? Doesn't really make much sense to me.
[See docs](https://www.webcomponents.org/element/firebase/polymerfire/firebase-auth) for readOnly mention.

**Comment by sbeleidy** [comment]
Thank you for the prompt reply.

Thanks for clarifying about the anonymous button. It's not really clear that the flat one is flat.

Could you please do the following?
* Add the sign out feature of login-fire-button as a boolean called canSignOut and default it to false
* Rename login-fire-panel back to login-fire

Thanks for all the hard work you're doing for this.

File: login-fire-social.html (hunk 2 above)
Comment (masonlouchart) [reply login-fire-social.html:56]: You can try it on the demo page. If you sign in with a button you can sign out with any other button that has the same `app-name`. On the `login-fire-social` demo page you can even sign in with Google and sign out with the Twitter, Facebook or Anonymous button.

**Comment by bkawk** [comment]
I found another SVG on Wikimedia to be swapped out for the google I con that was having issues 
https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg the code from it is below.

`<svg id="Layer_1" data-name="Layer 1" xmlns="http://www.w3.org/2000/svg" viewBox="0 0 472 472"><title>Google_-G-_Logo</title><path d="M482.56,261.36a271,271,0,0,0-4.29-48.27H256v91.29H383a108.59,108.59,0,0,1-47.09,71.23v59.21h76.27C456.82,393.73,482.56,333.23,482.56,261.36Z" transform="translate(-20 -20)" style="fill:#4285f4"/><path d="M256,492c63.72,0,117.14-21.13,156.19-57.18l-76.27-59.21c-21.13,14.16-48.17,22.53-79.92,22.53-61.47,0-113.49-41.51-132.05-97.3H45.1V362A235.9,235.9,0,0,0,256,492Z" transform="translate(-20 -20)" style="fill:#34a853"/><path d="M123.95,300.84a139.55,139.55,0,0,1,0-89.68V150H45.1a236.33,236.33,0,0,0,0,212Z" transform="translate(-20 -20)" style="fill:#fbbc05"/><path d="M256,113.86c34.65,0,65.76,11.91,90.22,35.29l67.69-67.69C373,43.39,319.61,20,256,20A235.93,235.93,0,0,0,45.1,150L124,211.16C142.51,155.38,194.54,113.86,256,113.86Z" transform="translate(-20 -20)" style="fill:#ea4335"/><path d="M20,20H492V492H20V20Z" transform="translate(-20 -20)" style="fill:none"/></svg>`

**Comment by masonlouchart** [comment]
Well done @bkawk ! It works like a charm 💯

**Comment by sbeleidy** [comment]
Hey @LM450N, just making sure that's all the commits? Or are you still working on something else? 

Also, just to make sure we're all on the same page. I think this PR closes the following:
#16 #40 #44 #45

Did I miss anything?

**Comment by masonlouchart** [comment]
Yes, all commits are here. No work in progress.

I don't think the issue #40 is solved.

The `login-fire-button` can sign out but it takes the CSS style of the given provider. Maybe will have to create a `login-fire-close-button` element that will simply implement the bahavior without applying style accordingly of the provider. Or, we can remove the style from the `login-fire-button` that can sign out (only when it's in the "signed-in" state).

You didn't miss anything 💯

**Comment by sbeleidy** [comment]
What do you both think about adding a none provider which would be hidden if user is not signed in and have plain styling if they are and would sign out.

We can also expose the styling variables for it through #7.

**Comment by masonlouchart** [comment]
Maybe we could expose customizable variables and simply add an attribute "neutral" on the `login-fire-button` to apply a "neutral style" that we'll define. If its value is false (default value), the button applies the provider style. If the user provides custom variables, they are prioritized. To do so, the developer has working social button, can easily switch to a neutral button (for signing out from a different page for example) and even define its own style.

Moreover, I thought about the `login-fire-social`. When the user is signed in, if the element can sign out, it should keeps only one close button that is the which one the user used to sign in.

The advantage of the initial idea (to have a separate element) is it has only one state and can only sign out.


__Which one do you prefer?__
-   [ ] customizable variables + attribute
-   [ ] login-fire-close-button element

**Comment by sbeleidy** [comment]
You've mentioned the idea that the provider the user uses to sign in stays and maintains the styling a couple times but I still think it's misleading. You can sign in with Facebook but you don't sign out with or of Facebook. You sign out of my app. In my mind, when you're signed in the sign out button should always show neutral colors. What are your thoughts on this? 

The only use case I see for keeping the provider styling is if it was something like disconnecting Facebook from my account for this app. Assuming you could connect other providers as well. 

Assuming that sign out would switch to neutral, we could 1- show only 1 button for social when logged out
1- use login-fire-button as a logout button anywhere else with no provider given

In the case where we don't do can-sign-out, should login-fire-button be hidden when the user is signed in?

**Comment by masonlouchart** [comment]
OK for removing provider style from `login-fire-button` when the user is signed in.

I don't know for hiding the button that can't sign out when a user is signed in. I don't know use case where an app displays a button that can only sign in and does not redirect the user on another page or displays an other button/menu to sign out. It was the reason I made the `login-fire-button` _"sign-out-able"_ by default. The developer can hide the button by himself (with `hidden$="[[signedIn]]"`). Moreover, it could have an impact on the displaying of the page. With a good documentation the developer will be warned. But if you think it will be a nice feature, we can hide the button by default when a user is signed in.

**Comment by bkawk** [comment]
I think there should always be a log out button, I can imagine the trauma of logging into an app on a shared computer and not being able to log out. 
The color of the logout button I think should be in line with the app, not with the sign in provider.

**Comment by masonlouchart** [comment]
Maybe we should do not add more features on this PR, that were not available before? We can continue to discuss on each opened issues and make smaller PR.

Anyway, I vote for the solution proposed by @bkawk  for the sign out button style, which is to apply the app theme.

**Comment by sbeleidy** [comment]
@LM450N Thank you so much for this PR! I'll merge it and we can work on the rest on several branches and discuss on the relevant issues/PRs.

**Pull request merged**